validator/src/state/merkle.rs

```diff
@@ -262,7 +266,9 @@ impl MerkleDatabase {
                     parent_node.children.is_empty()
                 };
 
-                if remove_parent {
+                // there's no children to the parent node already in the tree, remove it from
+                // path_map if a new node doesn't have this as its parent
+                if remove_parent && !additions.contains(parent_address) {
```

Comment:
I believe  you can use `path_map.contains_key` here and not produce the `additions` set.

Reply:
Reason I didn't use path_map here is because it has path_by_tokens() of delete_items (seen at line 250).
At line 254 only the delete_items are removed from path_map, leaving behind nodes in delete_items path in it.
``while`` loop at line 256 is for removing all parent elements of delete_items which are not required to be in the tree anymore and if I use path_map.contains_key() here, this will be evaluated to TRUE for all parent nodes of delete_items.

Let's consider the case:
path_map has {``ea``, ``ea12``, ``ea1234``, ``ea1278``, ``ea123456``, ``ea127890``}.
``ea127890`` is requested for addition.
``ea123456`` is requested for deletion, then ``ea123456`` will be removed from path_map at line 254.
However ``ea1234``, one of the parent node's path to be removed will be present in path_map and path_map.contains_key('``ea1234``') will be TRUE resulting in breaking this while loop at line 256.

To use path_map.contains_key(), line 250 shall be changed to not include any parent path of delete_items which has either single child or all of its children in path to delete_items.

Please correct me if I missed something here.

Reply:
I think you are correct, but I would perhaps figure out a way to get rid of that clone - it seems inefficient

Reply:
Though, through some tests of my own, this might be difficult with the current code.